dissect/target/plugins/os/windows/regf/usb.py

```diff
@@ -35,127 +44,173 @@
-
-        .. code-block:: text
-
-            hostname (string): The target hostname
-            domain (string): The target domain
-            type (string): Type of USB device
-            serial (string): Serial number of USB storage device
-            vid (string): Vendor ID of USB storage device
-            pid (string): Product ID of the USB storage device
-            rev (string): Version of the USB storage device
-            containerid (string):
-            friendlyname (string): Display name of the USB storage device
-            first_insert (datetime): First insertion date of USB storage device
-            first_install (datetime): First instalation date of USB storage device
-            last_insert (datetime): Most recent insertion (arrival) date of USB storage device
-            last_removal (datetime): Most recent removal (unplug) date of USB storage device
-            info_origin (string): Location of info present in output
+    @export(record=USBRegistryRecord)
+    def usb(self) -> Iterator[USBRegistryRecord]:
+        """Yields information about (historically) attached USB storage devices on Windows.
+
+        Uses the registry to find information about USB storage devices that have been attached to the system.
+        Also tries to find the past volume name and mount letters of the USB device and what user(s) interacted
+        with them using ``explorer.exe``.
         """
 
         for k in self.target.registry.keys(self.USB_STOR):
-            info_origin = "\\".join((k.path, k.name))
-            usb_stor = k.subkeys()
-
-            for usb_type in usb_stor:
-                device_info = self.parse_device_name(usb_type.name)
-                usb_devices = usb_type.subkeys()
-                for usb_device in usb_devices:
-                    properties = list(usb_device.subkeys())
+            for usb_type in k.subkeys():
+                try:
+                    device_info = parse_device_name(usb_type.name)
+                except ValueError:
+                    self.target.log.warning(f"Unable to parse USB device name: {usb_type.name}")
+                    device_info = {"type": None, "vendor": None, "product": None, "revision": None}
+
+                for usb_device in usb_type.subkeys():
                     serial = usb_device.name
+                    friendly_name = None
+                    container_id = None
+                    timestamps = {
+                        "first_install": None,
+                        "first_insert": None,
+                        "last_insert": None,
+                        "last_removal": None,
+                    }
+
+                    values = [v.name for v in usb_device.values()]
+                    if "FriendlyName" in values:
+                        friendly_name = usb_device.value("FriendlyName").value
+
+                    if "ContainerID" in values:
+                        container_id = usb_device.value("ContainerID").value
```

Comment:
I know it's not ideal but I'd prefer a try/except here for retrieving these values instead of this. Perhaps in the future we can improve the registry abstraction layer to have something similar to `dict.get(key, default=...)`.

Reply:
That could be nice, but a lot more work as well :smile: I've created a ticket for it: #778